accepted/0000-registry-spec.md

```diff
@@ -0,0 +1,160 @@
+# The `registry:` Dependency Specifier
+
+## Summary
+
+Add a dependency specifier which defines a registry base url, package name,
+and optionally SemVer range or dist-tag.
+
+## Motivation
+
+Occasionally, users wish to use multiple npm registries.  For example, they
+may have some packages hosted on the public npm registry, and others within
+a private registry on their company's intranet, or provided by a company
+like GitHub, Jfrog, Sonatype, or others.
+
+Currently, it is possible to map a scope to a given registry, and all
+packages starting with that scope will be published to and installed from
+the defined registry:
+
+```ini
+; .npmrc file
+@company:registry = https://npm-registry.my-company.com
+```
+
+However, this does not address the following use cases:
+
+- Users publish a package to one registry as a "staging" area, and then
+  later wish to "upgrade" it to the public npm registry for broader
+  distribution.  Some code should fetch from the internal registry (for
+  example, for testing and validation prior to promotion), but this is
+  challenging unless the private registry also can serve _any_ public npm
+  package.
+
+- Users have a set of unscoped package dependencies, some of which come
+  from the public registry, and others which have patches applied to them
+  (either to the code, or to the packument to add warnings via the
+  `deprecated` field for example).  Again, this is only feasible right now
+  by making the internal registry capable of proxying the entire set of npm
+  packages.
+
+- Alias package specifiers cannot point to any registries other than the
+  primary `--registry` configuration.  It would be useful in many scenarios
+  to be able to alias a package to a copy found on a different registry.
+
+- Migrating packages from one registry to another is particularly
+  challenging, requiring downloading the tarball locally and then
+  re-uploading it.  It would be much simpler to script such migrations by
+  being able to do `npm publish <pkg from src registry> --registry=<dest
+  registry>`.
+
+- Using multiple different registries in the best of cases always requires
+  using a `.npmrc` file for configuration, which is a challenge in some
+  environments, and always adds a layer of indirection.  It would be more
+  straightforward at times to specify the registry right in the
+  package.json file or command line.
+
+## Detailed Explanation
+
+A new dependency specifier is added:
+
+```
+registry:<registry url>#<package name>[@<specifier>]
```

Comment:
why not leveraging the current syntax? The `<source>:` syntax which is already used for github and some other registries.

if there are multiple packages from the same registry, the URL will be duplicated? Maybe it should be moved out to a separate field.

For instance:

```json
"dependencies": {
  "foo": "corp:^1.0.0",
  "bar": "corp:^1.0.0"
},
"registries": {
  "corp": "https://registry.corp.com"
}
```

also, if such a package will be published to the public registry, are all the third-party registries trusted by default?

Reply:
Defining a spec name to point to a registry is a good idea, and there's a number of ways we could go with various different trade-offs.  Worth doing as a subsequent RFC that builds on this one, since `corp:foo@1.x` would presumably desugar to `registry:https://registry.corp.com#foo@1.x`.

Reply:
It's not only a matter of syntactic sugar - if we support / implement this, then we need to be sure the syntax meet our standards. If fully qualified URLs lead to subpar developer experience, then it'll be very confusing to later say "our bad, now you can use this other syntax that work better but isn't as well supported".

Reply:
I don't really understand the objection here.  If we want to support a custom registry url in the package manifest as suggested here, how is that made any more difficult by _also_ supporting a full registry url as part of the dependency spec?  In fact, it seems like it would be somewhat easier implementation-wise, because most of the code that consumes specs would be able to remain agnostic as to whether the alias spec was `corp:foo@1.x` or `npm:foo@1.x` or `registry:https://registry.npmjs.org/#foo@1.x`, and we would have a verbose canonical way to save it that doesn't rely on having the rest of the package.json file in order to parse it.

Reply:
> If we want to support a custom registry url in the package manifest as suggested here, how is that made any more difficult by *also* supporting a full registry url as part of the dependency spec?

I think our point is that we're not convinced we want to support both, since that has a cost in terms of documentation, and has the potential to be confusing for our users. Should they use an url or a name? In which context? Etc. I'd much prefer having a single consistent syntax, and that makes it important to discuss what this syntax would be.

> we would have a verbose canonical way to save it that doesn't rely on having the rest of the package.json file in order to parse it

Imo registry names shouldn't be mapped to urls via the package.json, but rather by our respective configuration files - just like scope urls.